### ISSUE TITLE
Problem container-izing builds with Spack-built compilers, what's the good practice?

### ISSUE THREAD

**Issue by hartzell** [LLNL/spack#17273]
TL; DR, when I built a Docker image w/ bcl2fastq, compiled with a Spack-built gcc@8.4.0 starting from the Container docs page's `spack.yaml`, running the result fails for lack of various compiler libraries.

As a hack, including gcc@8.4.0 in the list of packages (and rebuilding it) works.

---

My general pattern with Spack is to build a gcc (usually @8.4.0 these days) and then use that to compile all of the things that I need.

I've tried to take that approach as I learn to use `spack containerize` (Docker, at the moment) things.  I've attempted a small optimization by extending the centos-7.dockerfile so that instead of running `spack speck hdf5+mpi` (what's the reason for that, anyway?) it:

- adds a packages.yaml file that prefers gcc@8.4.0;
- builds gcc@8.40; and
- adds it to the compiler list

but I think I'd see the same behavior if I just built the compiler as the first step of the container build.

I started from the [Container docs example spack.yaml](https://spack.readthedocs.io/en/latest/containers.html#environment-configuration), swapped the packages for `bcl2fastq` and commented out the bit that includes `libgomp`.

Turns out that that fails to run because it *needs* `libgomp` (funny that the package that I was practicing with also uses it...).  In my normal tree it picks up the library from within the gcc@8.4.0 package install.

I re-enabled loading `libcomp` as an external package.  The resulting image failed to run due to C (or C++) libraries not being the correct version.

My "solution" is to build and install both `bcl2fastq2` and `gcc@8.4.0` in the image, the compiler bits get linked into `/opt/view` and the various environment variables that are set seem to be sufficient to allow the various libraries to be found.

It won't be the end of the world if I have to include the compiler in every image, but I will be sad if I need to rebuild it (so that it can land in `/opt/software`) every time.

Are folks that are using `containerize` just using the OS compilers or ???

**Thanks!**

**Comment by hartzell** [comment]
ping @alalazo

**Comment by alalazo** [comment]
@hartzell Did you try using #15028 and:

- Build your custom Spack enabled base image that has `gcc@8.4.0` installed already
- Use that in the build stage

?

**Comment by hartzell** [comment]
I have not (*yet*) tried #15028, wasn't sure of it's status and wanted to see how dev behaved (that's what my cow-orkers would be likely to grab off the shelf).

I built my base image by changing a few lines at the bottom of the stock spack centos base image Dockerfile and I "used" it by editing the From line in the generated Dockerfile.

The problem seems to be that the gcc@8.4.0 I built with the spack that's in the base image is installed into `.../spack/opt/spack/...` while the things that are built for (and copied into) the final image land in `/opt/software`.

Can I customize the spack in the base image to build/install `gcc@8.4.0` into `/opt/software` **and** have it be the compiler used to build the final image packages by adding it via `spack compiler find`?

Seems like that's work; I'll try after my next meeting (*and* will get around to trying the PR).

Thanks for the followup!

**Comment by alalazo** [comment]
> Can I customize the spack in the base image to build/install gcc@8.4.0 into /opt/software and have it be the compiler used to build the final image packages by adding it via spack compiler find?

I didn't try that myself, but I think if you copy/create in your image a `config.yaml` like:
```yaml
config:
  install_tree: /opt/software
```
and then from the same directory you do:
```console
spack -C . install gcc@8.4.0
```
that should be installed in the right place.

**Comment by hartzell** [comment]
That seems to work, or more specifically this does:

```diff
--- share/spack/docker/my-centos-7.dockerfile	2020-06-26 22:50:17.023096008 +0000
+++ share/spack/docker/centos-7.dockerfile	2020-06-24 22:28:17.445059904 +0000
@@ -65,12 +65,8 @@
 WORKDIR /root
 SHELL ["docker-shell"]

-RUN ( echo '---' \
- &&   echo 'config:' \
- &&   echo '  install_tree: /opt/software' ) > $SPACK_ROOT/etc/spack/config.yaml
-COPY packages.yaml $SPACK_ROOT/etc/spack
-RUN spack install gcc@8.4.0
-RUN spack compiler add $(spack location -i gcc@8.4.0)
+# TODO: add a command to Spack that (re)creates the package cache
+RUN spack spec hdf5+mpi

 ENTRYPOINT ["/bin/bash", "/opt/spack/share/spack/docker/entrypoint.bash"]
 CMD ["interactive-shell"]
```

It ends up adding about 1.2GB to the resulting Docker images.  I can cut that down to 1.1GB by running `spack gc` in the base image.  I can knock that down to 247MB if I remove all things in the gcc dir except `lib` and `lib64`, which appears to still leave me with a functional application.

**Comment by hartzell** [comment]
What's the future look like for #15028?

**Comment by alalazo** [comment]
@hartzell I hope it's almost in its final stage. Reviewing process has been slow due to other priorities so far, but with that I was able to e.g. produce a minimal size image of `gromacs+cuda` see https://hub.docker.com/repository/docker/alalazo/gromacs/tags?page=1

**Comment by hartzell** [comment]
I'm happy with the solution that I've arrived at and am looking forward to rolling it into the changes in #15028 so I'm closing this.
